File: usvm-ts/src/test/kotlin/org/usvm/samples/FieldAccess.kt
package org.usvm.samples

import org.jacodb.ets.model.EtsScene
import org.junit.jupiter.api.Disabled
import org.usvm.api.TsValue
import org.usvm.util.TsMethodTestRunner
import kotlin.test.Test

class FieldAccess : TsMethodTestRunner() {

    private val className = this::class.simpleName!!

    override val scene: EtsScene = loadSampleScene(className)

    @Test
    fun `test readDefaultField`() {
        val method = getMethod(className, "readDefaultField")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 5.0 },
        )
    }

    @Test
    fun `test writeAndReadNumeric`() {
        val method = getMethod(className, "writeAndReadNumeric")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 14.0 },
        )
    }

    @Test
    fun `test writeDifferentTypes`() {
        val method = getMethod(className, "writeDifferentTypes")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 5.0 },
        )
    }

    @Test
    fun `test handleNumericEdges`() {
        val method = getMethod(className, "handleNumericEdges")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 1.0 },
        )
    }

    @Test
    fun `test createWithField`() {
        val method = getMethod(className, "createWithField")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 15.0 },
        )
    }

    @Disabled("Return types are not propagated to locals, need type stream")
    @Test
    fun `test factoryCreatedObject`() {
        val method = getMethod(className, "factoryCreatedObject")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 42.0 },
        )
    }

    @Test
    fun `test conditionalFieldAccess`() {
        val method = getMethod(className, "conditionalFieldAccess")
        discoverProperties<TsValue.TsClass, TsValue.TsNumber>(
            method = method,
            { a, r ->
                val x = a.properties["x"] as TsValue.TsNumber
                x.number == 1.1 && r.number == 14.0
            },
            { a, r ->
                val x = a.properties["x"] as TsValue.TsNumber
                x.number != 1.1 && r.number == 10.0
            },
        )
    }

    @Disabled("Nested field types are not propagated to locals, need type stream")
    @Test
    fun `test nestedFieldAccess`() {
        val method = getMethod(className, "nestedFieldAccess")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 7.0 },
        )
    }

    @Disabled("Nested arrays inside objects are accessed via field properties ('.1') instead of indices ([1])")
    @Test
    fun `test arrayFieldAccess`() {
        val method = getMethod(className, "arrayFieldAccess")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 5.0 },
        )
    }

    @Test
    fun `test multipleFieldInteraction`() {
        val method = getMethod(className, "multipleFieldInteraction")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 9.0 }, // (2*2=4) + (4+1=5) == 9
        )
    }

    @Test
    fun `test circularTypeChanges`() {
        val method = getMethod(className, "circularTypeChanges")
        discoverProperties<TsValue.TsNumber>(
            method = method,
            { r -> r.number == 1.0 },
        )
    }
}

File: usvm-ts/src/test/resources/samples/lang/FieldAccess.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols

class FieldAccess {
    readDefaultField(): number {
        const obj = new SimpleClass();
        return obj.x;
    }

    writeAndReadNumeric(): number {
        const obj = new SimpleClass();
        obj.x = 14;
        return obj.x;
    }

    writeDifferentTypes(): number {
        const obj = new SimpleClass();
        obj.x = true;
        if (obj.x === true) {
            obj.x = undefined;
        }
        if (obj.x === undefined) {
            obj.x = null;
        }
        if (obj.x === null) {
            obj.x = 5;
        }
        return obj.x;
    }

    handleNumericEdges(): number {
        const obj = new SimpleClass();
        obj.x = NaN;
        if (Number.isNaN(obj.x)) {
            obj.x = Infinity;
        }
        return (obj.x === Infinity) ? 1 : 0;
    }

    createWithField(): number {
        return { a: 15 }.a;
    }

    factoryCreatedObject(): number {
        return this.createObject().x;
    }

    conditionalFieldAccess(a: SimpleClass): number {
        if (a.x === 1.1) return 1;
        return 2;
    }

    nestedFieldAccess(): number {
        const obj = { inner: new SimpleClass() };
        obj.inner.x = 7;
        return obj.inner.x;
    }

    arrayFieldAccess(): number {
        const obj = { arr: [7, 3, 6] };
        obj.arr[1] = 5;
        return obj.arr[1];
    }

    multipleFieldInteraction(): number {
        const obj = new MultiFieldClass();
        obj.a = obj.b * 2;
        obj.b = obj.a + 1;
        return obj.a + obj.b;
    }

    circularTypeChanges(): number {
        const obj = new SimpleClass();
        obj.x = { value: 5 };
        obj.x = obj.x.value;
        obj.x = [obj.x];
        return obj.x.length;
    }

    private createObject(): { x: number } {
        return { x: 42 };
    }

    readFromAny(x: any): number {
        if (x === undefined) return 0;
        if (x === null) return 0;
        if (x.a === 10) return 1;
        return 2;
    }

    writeToAny(x: any): number {
        if (x === undefined) return 0;
        if (x === null) return 0;
        x.a = 20;
        x.b = 30;
        return x.a + x.b;
    }
}

class SimpleClass {
    x: any = 5;
}

class MultiFieldClass {
    a = 1;
    b = 2;
}
